browser/src/Services/Git.ts

```diff
@@ -5,28 +5,109 @@
  */
 
 import { exec } from "child_process"
+import * as path from "path"
+import * as git from "simple-git/promise"
+import { promisify } from "util"
+import * as Log from "./../Log"
+
+// FIXME: Returns a string however is typed to return { stderr: any; stdout: any }
+const execPromise = promisify(exec) as any
 
 interface IExecOptions {
     cwd?: string
 }
 
-export function getBranch(path?: string): Promise<string> {
-    return new Promise((resolve, reject) => {
-        const options: IExecOptions = {}
-        if (path) {
-            options.cwd = path
+interface IStatus extends git.DiffResult {
+    modified?: number
+}
+
+export interface GitFunctions {
+    getGitSummary(workspace: string): Promise<git.DiffResult | null>
+    getBranch(path?: string): Promise<Error | string>
+    getGitRoot(): Promise<string | null>
+}
+
+const numFromString = (s: string) => {
+    const num = s.match(/\d+/)
+    if (num && num.length) {
+        return Number(num[0])
+    }
+    return null
+}
+
+const formatFileAndChanges = (files: string[]) => {
+    if (!files.length) {
+        return null
+    }
+    return files.map(unformattedStr => {
+        const [file, changes] = unformattedStr.split("|")
+        const pluses = changes.replace(/[^+]/g, "")
+        const insertions = pluses.length
+        const minuses = changes.replace(/[^-]/g, "")
+        const deletions = minuses.length
+        return {
+            file: file.trim(),
+            changes: numFromString(changes),
+            insertions,
+            deletions,
         }
+    })
+}
 
-        exec(
-            "git rev-parse --abbrev-ref HEAD",
-            options,
-            (error: any, stdout: string, stderr: string) => {
-                if (error && error.code) {
-                    reject(new Error(stderr))
-                } else {
-                    resolve(stdout)
+export async function getGitRoot(): Promise<string | null> {
+    try {
+        const rootDir = await (git() as any).revparse(["--show-toplevel"])
+        Log.info(`Git Root Directory is ${rootDir}`)
+        return path.join(__dirname, rootDir)
+    } catch (e) {
+        return null
+    }
+}
+
+export async function getGitSummary(currentDir: string): Promise<IStatus | null> {
+    let status = null
+    if (currentDir) {
+        // TODO: .customBinary()
+        const project = git(currentDir)
+        const isRepo = await project.checkIsRepo()
+        if (isRepo) {
+            // status = await project.diffSummary()
+            // if (status.files || status.deletions || status.insertions) {
+            //     return status
+            // }
+            const options: IExecOptions = {
+                cwd: currentDir,
+            }
+            const cmd = `git diff --stat=4096`
+            try {
```

Comment:
Following up from your issue https://github.com/steveukx/git-js/issues/241 - this block should be equivalent of using:

```
status = await project.diffSummary();
```

Do you find that the `status` is an instance of the `simple-git` object rather than an instance of `DiffSummary` when run this way?

Reply:
@steveukx the bug I'm currently facing is pretty odd and I'm not sure it isn't electron related. Basically when I run `await project.diffSummary()` **in development** it works perfectly for me though when I build and package the electron app the same variable `status` now returns the `simple-git` object so my assumption is somehow something changes once packaged.

Except when I run the raw command via `childProcess` that outputs the git command 😕, haven't found reference to a similar issue anywhere so i'm stumped

Reply:
That is strange - is it certain that the library is included as `require('simple-git/promise')` ? if it's included as `require('simple-git')` you would get the `simple-git` object back to chain it on to additional commands.

Reply:
Yes I've `imported` it as `git/promise` tried `simple-git` but switched over once I realised we would be able to use promises